### PR TITLE
[SourceKit] Pass `ResolvedCursorInfo` as shared pointer instead of by value

### PR DESCRIPTION
**Pull request by ahoppen**
As discussed in https://github.com/apple/swift/pull/62292#discussion_r1035212344.

This allows us to model the `ResolvedCursorInfo` types as a proper type hierarchy instead of having to store all values in the base `ResolvedCursorInfo` type.

rdar://102853071

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by bnbarham** [comment]
I haven't looked through it all, but is there some reason this can't be a `unique_ptr`? I would have thought the ownership wouldn't be that complicated here.

File: lib/IDE/IDERequests.cpp [hunk 1]
```diff
@@ -392,13 +395,13 @@ void swift::simple_display(llvm::raw_ostream &out, const CursorInfoOwner &owner)
 }
 
 void swift::ide::simple_display(llvm::raw_ostream &out,
-                                const ide::ResolvedCursorInfo &info) {
-  if (info.isInvalid())
+                                ide::ResolvedCursorInfoPtr info) {
```
Comment (bnbarham) [inline lib/IDE/IDERequests.cpp:398]: In general I'd prefer things that were being passed a const ref to remain that way (especially for methods like this).

**Comment by bnbarham** [comment]
> I haven't looked through it all, but is there some reason this can't be a unique_ptr? I would have thought the ownership wouldn't be that complicated here.

Ah. We cache the result, so we pretty much need to use a shared pointer here. Fair enough.

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 2]
```diff
@@ -2446,12 +2447,12 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
...
-          evaluateOrDefault(SrcFile.getASTContext().evaluator,
-                            CursorInfoRequest{CursorInfoOwner(&SrcFile, Loc)},
-                            ResolvedCursorInfo());
+        ResolvedCursorInfoPtr CursorInfo =
+            evaluateOrDefault(SrcFile.getASTContext().evaluator,
+                              CursorInfoRequest{CursorInfoOwner(&SrcFile, Loc)},
+                              new ResolvedCursorInfo());
         auto ValueRefCursorInfo =
-            dyn_cast<ResolvedValueRefCursorInfo>(&CursorInfo);
+            dyn_cast_cursor_info<ResolvedValueRefCursorInfo>(CursorInfo);
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2455]: `dyn_cast` can take an intrusive pointer and returns a `*`. I don't think we need `dyn_cast_cursor_info`.

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 3]
```diff
@@ -2446,12 +2447,12 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
...
           return;
 
-        ResolvedCursorInfo CursorInfo =
-          evaluateOrDefault(SrcFile.getASTContext().evaluator,
-                            CursorInfoRequest{CursorInfoOwner(&SrcFile, Loc)},
-                            ResolvedCursorInfo());
+        ResolvedCursorInfoPtr CursorInfo =
+            evaluateOrDefault(SrcFile.getASTContext().evaluator,
+                              CursorInfoRequest{CursorInfoOwner(&SrcFile, Loc)},
+                              new ResolvedCursorInfo());
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2453]: I'd prefer this just be `nullptr` and remove the invalid state of `ResolvedCursorInfo`. Happy for that to be in a different PR though.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 4]
```diff
@@ -1160,15 +1160,15 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
 
 /// Returns true on success, false on error (and sets `Diagnostic` accordingly).
 static bool passCursorInfoForDecl(
-    const ResolvedValueRefCursorInfo &Info, bool AddRefactorings,
+    ResolvedValueRefCursorInfoPtr Info, bool AddRefactorings,
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:1163]: This is the only use of `ResolvedValueRefCursorInfoPtr` right? It could theoretically stay as a `const ResolvedValueRefCursorInfo &Info` and then just stack allocate `ResolvedValueRefCursorInfo` in the two callers. But also I strongly dislike all the callbacks here, so maybe this will make it easier to get rid of that later 🤷

**Review by bnbarham — APPROVED** [review]

File: lib/IDE/CursorInfo.cpp [hunk 5]
```diff
@@ -411,7 +413,7 @@ class CursorInfoDoneParsingCallback : public IDEInspectionCallbacks {
       break;
     }
     if (Result) {
-      Consumer.handleResults(*CursorInfo);
+      Consumer.handleResults(CursorInfo);
     }
```
Comment (bnbarham) [inline lib/IDE/CursorInfo.cpp:417]: `CursorInfo` could have been null here before 😅. I'm honestly not sure why it doesn't actually cause any issues though. It does on Linux with C++ stdlib assertions enabled. Anyway, looks like this fixes it (assuming `handleResults` handles `nullptr`).

**Comment by ahoppen** [comment]
@swift-ci Please smoke test macOS
